scripts/migration/migrate-istio

```diff
@@ -0,0 +1,1360 @@
+
+ASM requires you to have at least ${TOTAL_CPU_REQ} vCPUs in node pools whose
+machine type is at least ${MACHINE_CPU_REQ} vCPUs.
+${CLUSTER_LOCATION}/${CLUSTER_NAME} does not meet this requirement. ASM
+may not function as expected.
+
+EOF
+  fi
+}
+
+validate_istio_version() {
+  info "Checking existing Istio version(s)..."
+  local VERSION_OUTPUT; VERSION_OUTPUT="$(retry 2 istioctl version -o json)"
+  if [[ -z "${VERSION_OUTPUT}" ]]; then
+    fatal "Couldn't validate existing Istio versions."
+  fi
+  local FOUND_INVALID_VERSION; FOUND_INVALID_VERSION=0
+  for version in $(echo "${VERSION_OUTPUT}" | jq -r '.meshVersion[].Info.version' -r); do
+    if [[ "$version" =~ ^1.7 || "$version" =~ ^1.8 || "$version" =~ ^1.9 || "$version" =~ ^1.10 ]]; then
+      info "  $version"
+    else
+      info "  $version (not supported)"
+      FOUND_INVALID_VERSION=1
+    fi
+  done
+  if [[ "$FOUND_INVALID_VERSION" -eq 1 ]]; then
+    fatal "Migration requires existing control planes at least version 1.7."
+  fi
+  info "No version issues found."
+}
+
+ensure_backup() {
+  if ! [[ -d ./configuration_backup ]]; then
+    fatal "Configuration backup not found, please run pre-migration step and retry installation."
+  fi
+}
+
+configure_package() {
+  info "Configuring kpt package..."
+
+  run kpt cfg set asm gcloud.container.cluster "${CLUSTER_NAME}"
+  run kpt cfg set asm gcloud.core.project "${PROJECT_ID}"
+  run kpt cfg set asm gcloud.project.environProjectNumber "${FLEET_PROJECT_NUMBER}"
+  run kpt cfg set asm gcloud.compute.location "${CLUSTER_LOCATION}"
+  run kpt cfg set asm gcloud.compute.network "${GCE_NETWORK_NAME}"
+  run kpt cfg set asm anthos.servicemesh.rev "${REVISION_LABEL}"
+  run kpt cfg set asm anthos.servicemesh.tag "${ASM_VERSION}"
+  run kpt cfg set asm anthos.servicemesh.hubTrustDomain "${FLEET_PROJECT_ID}.svc.id.goog"
+  HUB_IDP_URL="$(kubectl get memberships.hub.gke.io membership -o=jsonpath='{.spec.identity_provider}' || true)"
+  run kpt cfg set asm anthos.servicemesh.hub-idp-url "${HUB_IDP_URL}"
+  run kpt cfg set asm anthos.servicemesh.trustDomainAliases \
+    "${FLEET_PROJECT_ID}.svc.id.goog" "${FLEET_PROJECT_ID}.hub.id.goog"
+  info "Configured."
+}
+
+install_asm() {
+  local PARAMS
+  PARAMS="-f ${OPERATOR_MANIFEST}"
+  PARAMS="${PARAMS} -f ${CNI_MANIFEST}"
+  if [[ "${UPGRADE_GATEWAYS}" -eq 0 ]]; then
+    PARAMS="${PARAMS} -f ${NO_DEFAULT_INGRESS}"
+  fi
+  PARAMS="${PARAMS} -c ${KUBECONFIG}"
+  PARAMS="${PARAMS} --set revision=${REVISION_LABEL}"
+  PARAMS="${PARAMS} --skip-confirmation"
+
+  info "Installing ASM control plane..."
+  # shellcheck disable=SC2086
+  retry 2 run istioctl install $PARAMS
+
+  # Prevent the stderr buffer from ^ messing up the terminal output below
+  sleep 1
+
+  local RAW_YAML; RAW_YAML="${REVISION_LABEL}-manifest-raw.yaml"
+  local EXPANDED_YAML; EXPANDED_YAML="${REVISION_LABEL}-manifest-expanded.yaml"
+
+  PARAMS="-f ${OPERATOR_MANIFEST}"
+  if [[ "${UPGRADE_GATEWAYS}" -eq 0 ]]; then
```

Comment:
nit: In the help strings, should we say that UPGRADE_GATEWAYS (default option) will result in a new IP?

Reply:
I don't think it should result in a new IP since the service isn't changing. It didn't happen to me during testing. If it happened to you let me know since that's pretty unexpected.